vshard/service_info.lua

```diff
@@ -0,0 +1,181 @@
+-- vshard.service_info
+--
+-- The preliminary purpose of the module is to ease the testing of the
+-- background fibers using luatest. It's also used for monitoring of the
+-- vshard's services and shows whether any errors occurred and what this
+-- service is doing right now.
+--
+
+local consts = require('vshard.consts')
+local rlist = require('vshard.rlist')
+local util = require('vshard.util')
+local fiber_clock = require('fiber').clock
+local gsc = util.generate_self_checker
+
+--
+-- Helper function to get the number of seconds, minutes or hours
+-- that past since the time.
+--
+local function time_past_since(time)
+    local time_past = math.floor(fiber_clock() - time)
+    local units = {'s', 'm', 'h'}
+    for i, unit in ipairs(units) do
+        if time_past < 60 or not next(units, i) then
+            return time_past, unit
+        end
+        time_past = math.floor(time_past / 60)
+    end
+end
+
+--------------------------------------------------------------------------------
+-- Service description
+--------------------------------------------------------------------------------
+
+local SERVICE_TEMPLATE = {
+    -- Intended to show whether everything is all right.
+    status = 'ok',
+    -- Increasing number of status changes.
+    status_idx = 0,
+    -- Intended to show what the service is doing right now.
+    activity = 'unknown',
+    -- List of errors that have occurred and have to be fixed.
+    errors = rlist.new(),
+}
+
+local function service_set_status(service, status)
+    service.status_idx = service.status_idx + 1
+    service.status = status
+end
+
+-- Just for consistency
+local function service_set_activity(service, activity)
+    service.activity = activity
+end
+
+--
+-- Saves an error if the previous was not the same one.
+--
+local function service_add_error(service, err)
+    local errors = service.errors
+    -- Drop the oldest error if needed
+    if errors.count >= consts.DEFAULT_SERVICE_INFO_ERRORS_NUMBER then
+        errors:remove(errors.first)
+    end
+    -- Add a new one
+    if not service.errors.last or service.errors.last.error ~= err then
```

Comment:
There seems to be a problem with which errors are considered the same:
```
tarantool> e1 = box.error.new(box.error.ILLEGAL_PARAMS, "bla bla")
---
...

tarantool> e2 = box.error.new(box.error.ILLEGAL_PARAMS, "bla bla")
---
...

tarantool> e1 == e2
---
- false
...
```
As you can see, they are the same (each pair of fields of these `struct error` objects is the same), but not equal, because the comparison is done by pointer. I would suggest that we should compare them by `code`, `type`, `message`. `trace` and `prev` we could probably skip, too complicated to compare them so deep.

Or otherwise I wouldn't mind adding same errors here multiple times. It is not like we are interested in old errors which are not actual anymore.

Another related comment is that are you sure we really need to store multiple errors at all? By the same reason - old errors are not really relevant anymore. Only the error happened during latest service work iteration is useful anyway.

If you are worrying about how a service would produce and save multiple errors during one iteration of its work loop - from experience usually the first error is the main one. Others are likely to be same or be triggered by the first one. So like in the RFC, each service usually looks like a loop:
```
while true do
    do_step1;
    do_step2;
    do_step3;
    ...
end
```
If an error happens on any of the steps, we save it and keep until next iteration, ignoring newer errors. Like that:
```
while true do
    first_err = nil
    if not do_step1 then
        if not first_err then
            first_err = err
            set_error(err)
        end
    if no do_step2 then
        same ...
    end
    if no do_step3 then
        same ...
    end
    ...
end
```
In your current approach multiple errors happening after the first one would just purge it out, if these errors are not the same. And we won't see the original first error of the current iteration.

I am not saying we need to redo it with just one error now. Only thoughts stream for now.

Reply:
> I am not saying we need to redo it with just one error now. Only thoughts stream for now.

Why not. Let's save only first error.